### PR TITLE
gh-105436： the environment block should end with two zero-valued wchar_t values

### PR DESCRIPTION
**Pull request by sku2000**
[In the CreateProcessW document, the environment block should end with two zero-valued wchar_t values（i.e. four zero bytes）](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa).

* Issue: gh-105436

### CODE REVIEW

File: Modules/_winapi.c [hunk 1]
```diff
@@ -796,6 +796,17 @@ getenvironment(PyObject* environment)
...
+
+    /* A Unicode environment block is terminated by four zero bytes:
+       two for the last string, two more to terminate the block. */
+    if (envsize == 0) {
+        buffer = PyMem_NEW(wchar_t, 2);
+        p = buffer;
+        *p++ = L'\0';
+        *p++ = L'\0';
+        goto error;
+    }
```
Comment (eryksun) [inline Modules/_winapi.c]: Using `goto error` is misleading. I suggest changing the function's exit label name to "cleanup".

Using `PyMem_Calloc()` avoids the need to manually zero the buffer for the null characters. Also, make sure to raise a `MemoryError` exception if allocation fails.

```suggestion
    if (envsize == 0) {
        // A environment block must be terminated by two null characters --
        // one for the last string and one for the block.
        buffer = PyMem_Calloc(2, sizeof(wchar_t));
        if (!buffer) {
            PyErr_NoMemory();
        }
        goto cleanup;
    }
```

Comment (sku2000) [reply Modules/_winapi.c]: thanks

File: Misc/NEWS.d/next/Windows/2023-06-08-11-30-17.gh-issue-105436.1qlDxw.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+the environment block should end with two zero-valued wchar_t values
```
Comment (eryksun) [inline Misc/NEWS.d/next/Windows/2023-06-08-11-30-17.gh-issue-105436.1qlDxw.rst]:
```suggestion
Ensure that an empty environment block is terminated by two null characters,
as is required by Windows.
```

**Comment by sku2000** [comment]
I can't reproduce the problem in the test.
This question may be related to this [issue].(https://github.com/python/cpython/issues/104791)

**Comment by zooba** [comment]
I guess the new test has exposed a leak on Ubuntu that no other tests trigger.

Maybe make the new test Windows only and tag it with this issue? No reason to block your change here - it's definitely not the cause of that leak.

**Comment by eryksun** [comment]
Failing with `STATUS_DLL_NOT_FOUND` (0xC0000135, 3221225781) at startup is probably due to a DLL init routine trying to load another DLL from a path that's based on a standard environment variable, such as `SystemRoot`. There's really no need to test the exit status. Just assert that `OSError` isn't raised.

**Comment by zooba** [comment]
Are we copying `python.exe` for this test but not moving `python312.dll`? That would be causing everything to fail if so.

**Comment by eryksun** [comment]
There's already a `test_empty_env()` test, which is skipped on Windows because the `SystemRoot` environment variable is required to start Python prior to 3.11. For example, here's 3.10 failing to start:

```pycon
>>> sys.version_info[:2]
(3, 10)
>>> subprocess.call(sys.executable, env={'a':'b'})
Fatal Python error: _Py_HashRandomization_Init: failed to get random numbers to initialize Python
Python runtime state: preinitialized
```

The `SystemRoot` environment variable is no longer required to start Python 3.11+, at least not on x86/x64, because the internal `win32_urandom()` function was updated to call `BCryptGenRandom()` instead of `CryptAcquireContextW()` / `CryptGenRandom()`.

The bot that's failing is an ARM64 system. I don't know whether process initialization  under ARM64 depends on `SystemRoot` or some other environment variable in order to load a library that's particular to the API implementation on ARM64. What I do know is that for this Windows-only test, it doesn't matter whether the child process can start and exit with some know exit status. All we need to check is that the `CreateProcessW()` call doesn't fail, meaning that in principle it's possible to spawn a child process with an empty environment.

**Pull request merged**